lib/MultiCompiler.js

```diff
@@ -2,175 +2,162 @@
 	MIT License http://www.opensource.org/licenses/mit-license.php
 	Author Tobias Koppers @sokra
 */
-var Tapable = require("tapable");
-var asyncLib = require("async");
-var MultiWatching = require("./MultiWatching");
-var MultiStats = require("./MultiStats");
+"use strict";
 
-function MultiCompiler(compilers) {
-	Tapable.call(this);
-	if(!Array.isArray(compilers)) {
-		compilers = Object.keys(compilers).map(function(name) {
-			compilers[name].name = name;
-			return compilers[name];
+const Tapable = require("tapable");
+const asyncLib = require("async");
+const MultiWatching = require("./MultiWatching");
+const MultiStats = require("./MultiStats");
+
+module.exports = class MultiCompiler extends Tapable {
+	constructor(compilers) {
+		super();
+		if(!Array.isArray(compilers)) {
+			compilers = Object.keys(compilers).map((name) => {
+				compilers[name].name = name;
+				return compilers[name];
+			});
+		}
+		this.compilers = compilers;
+		this.delegateProperty.call(this, "outputFileSystem");
+		this.delegateProperty.call(this, "inputFileSystem");
+		Object.defineProperty(this, "outputPath", {
+			configurable: false,
+			get: () => {
+				let commonPath = compilers[0].outputPath;
+				for(let i = 1; i < compilers.length; i++) {
+					while(compilers[i].outputPath.indexOf(commonPath) !== 0 && /[\/\\]/.test(commonPath)) {
+						commonPath = commonPath.replace(/[\/\\][^\/\\]*$/, "");
+					}
+				}
+				if(!commonPath && compilers[0].outputPath[0] === "/") return "/";
+				return commonPath;
+			}
 		});
+
+		let doneCompilers = 0;
+		let compilerStats = [];
+		this.compilers.forEach((compiler, idx) => {
+			let compilerDone = false;
+			compiler.plugin("done", (stats => {
+				if(!compilerDone) {
+					compilerDone = true;
+					doneCompilers++;
+				}
+				compilerStats[idx] = stats;
+				if(doneCompilers === this.compilers.length) {
+					this.applyPlugins("done", new MultiStats(compilerStats));
+				}
+			}));
+			compiler.plugin("invalid", (() => {
+				if(compilerDone) {
+					compilerDone = false;
+					doneCompilers--;
+				}
+				this.applyPlugins("invalid");
+			}));
+		}, this);
 	}
-	this.compilers = compilers;
 
-	function delegateProperty(name) {
+	delegateProperty(name) {
 		Object.defineProperty(this, name, {
```

Comment:
It would be great to refactor all these defineProperty calls to getter and setter

``` js
get outputFilesystem() {
  return ...
}

set outputFilesystem() {
  ...
}
```